libbeat/beat/beat.go

```diff
@@ -150,70 +139,113 @@ func init() {
+	return handleError(newBeat(name, version).launch(bt))
+}
+
+// newBeat creates a new beat instance
+func newBeat(name, version string) *Beat {
+	return &Beat{
+		Name:    name,
+		Version: version,
+		UUID:    uuid.NewV4(),
+	}
+}
+
+func (b *Beat) launch(bt BeatCreator) error {
+	if err := b.handleFlags(); err != nil {
+		return err
+	}
+
+	svc.BeforeRun()
+	defer svc.Cleanup()
+
+	if err := b.configure(); err != nil {
+		return err
+	}
+
+	logp.Info("Setup Beat: %s; Version: %s", b.Name, b.Version)
+	processors, err := processors.New(b.Config.Processors)
+	if err != nil {
+		return fmt.Errorf("error initializing processors: %v", err)
+	}
+
+	debugf("Initializing output plugins")
+	publisher, err := publisher.New(b.Name, b.Config.Output, b.Config.Shipper, processors)
+	if err != nil {
+		return fmt.Errorf("error initializing publisher: %v", err)
+	}
+
+	// TODO: some beats race on shutdown with publisher.Stop -> do not call Stop yet,
+	//       but refine publisher to disconnect clients on stop automatically
+	// defer publisher.Stop()
+
+	b.Publisher = publisher
+	beater, err := bt(b, b.RawConfig)
+	if err != nil {
+		return err
 	}
 
-	return &instance{
-		data: &Beat{
-			Name:    name,
-			Version: version,
-			UUID:    uuid.NewV4(),
-			BT:      bt,
-		},
-		beater: bt,
+	// If -configtest was specified, exit now prior to run.
+	if cfgfile.IsTestConfig() {
+		fmt.Println("Config OK")
+		return GracefulExit
 	}
+
+	svc.HandleSignals(beater.Stop)
+
+	logp.Info("%s start running.", b.Name)
+	defer logp.Info("%s stopped.", b.Name)
+
+	return beater.Run(b)
 }
 
 // handleFlags parses the command line flags. It handles the '-version' flag
 // and invokes the HandleFlags callback if implemented by the Beat.
-func (bc *instance) handleFlags() error {
+func (b *Beat) handleFlags() error {
 	// Due to a dependence upon the beat name, the default config file path
 	// must be updated prior to CLI flag handling.
-	err := cfgfile.ChangeDefaultCfgfileFlag(bc.data.Name)
+	err := cfgfile.ChangeDefaultCfgfileFlag(b.Name)
 	if err != nil {
 		return fmt.Errorf("failed to set default config file path: %v", err)
 	}
-
 	flag.Parse()
 
 	if *printVersion {
-		fmt.Printf("%s version %s (%s), libbeat %s\n", bc.data.Name,
-			bc.data.Version, runtime.GOARCH, defaultBeatVersion)
+		fmt.Printf("%s version %s (%s), libbeat %s\n",
+			b.Name, b.Version, runtime.GOARCH, defaultBeatVersion)
 		return GracefulExit
 	}
 
-	// Invoke HandleFlags if FlagsHandler is implemented.
-	if flagsHandler, ok := bc.beater.(FlagsHandler); ok {
-		err = flagsHandler.HandleFlags(bc.data)
-	}
-
-	return err
+	return handleFlags(b)
 }
 
 // config reads the configuration file from disk, parses the common options
 // defined in BeatConfig, initializes logging, and set GOMAXPROCS if defined
 // in the config. Lastly it invokes the Config method implemented by the beat.
-func (bc *instance) config() error {
+func (b *Beat) configure() error {
```

Comment:
I would prefer to keep the name config or if we change it, also comment must be changed.